src/gleam/http.gleam

```diff
@@ -106,6 +110,312 @@ pub fn method_from_dynamic(value: Dynamic) -> Result(Method, List(DecodeError))
   }
 }
 
+pub type MultipartHeaders {
+  MultipartHeaders(headers: List(Header), remaining: BitString)
+  MoreRequiredForHeaders(
+    continuation: fn(BitString) -> Result(MultipartHeaders, Nil),
+  )
+}
+
+pub type MultipartBody {
+  MultipartBody(chunk: String, done: Bool, remaining: BitString)
+  MoreRequiredForBody(
+    chunk: String,
+    continuation: fn(BitString) -> Result(MultipartBody, Nil),
+  )
+}
+
+/// Parse the headers for the next multipart part.
+///
+/// This function is used for parsing the multipart format is defined in
+/// RFC 2045, along with `parse_multipart_body`.
+///
+/// This function skips any preamble before the boundary. The preamble may be
+/// retrieved using `parse_body`.
+///
+/// This function will accept input of any size, it is up to the caller to limit
+/// it if needed.
+/// 
+pub fn parse_multipart_headers(
+  data: BitString,
+  boundary: String,
+) -> Result(MultipartHeaders, Nil) {
+  let boundary = bit_string.from_string(boundary)
+  // TODO: rewrite this to use a bit pattern once JavaScript supports
+  // the `b:binary-size(bsize)` pattern.
+  let prefix = <<45, 45, boundary:bit_string>>
+  case bit_string.slice(data, 0, bit_string.byte_size(prefix)) == Ok(prefix) {
+    // There is no preamble, parse the headers.
+    True -> parse_headers_after_prelude(data, boundary)
+    // There is a preamble, skip it before parsing.
+    False -> skip_preamble(data, boundary)
+  }
+}
+
+/// Parse the body of the current multipart part, which is everything until the
+/// next boundary.
+/// 
+/// This function is used for parsing the multipart format is defined in
+/// RFC 2045, along with `parse_multipart_headers`.
+///
+/// This function will accept input of any size, it is up to the caller to limit
+/// it if needed.
+/// 
+pub fn parse_multipart_body(
+  data: BitString,
+  boundary: String,
+) -> Result(MultipartBody, Nil) {
+  let boundary = bit_string.from_string(boundary)
+  parse_body_with_bit_string(data, boundary)
```

Comment:
nit: I might pipe this instead of redeclaring

(feel free to ignore all my code style suggestions, obviously)